### PR TITLE
Add ignore_permission_denied option

### PR DESCRIPTION
**Pull request by aminalaee**
Closes https://github.com/samuelcolvin/watchfiles/issues/147

### CODE REVIEW

**Comment by aminalaee** [comment]
I had two doubts about this:
- One is obviously the name
- Should it suppress all errors? Or should it ignore only permission denied, etc?

**Comment by samuelcolvin** [comment]
I'll review properly soon, but would be good if we respected an env variable so it could be set when using watchfiles with uvicorn etc. - that's much more common than the command line usage.

File: src/lib.rs [hunk 1]
```diff
@@ -45,36 +45,48 @@ struct RustNotify {
     watcher: WatcherEnum,
 }
 
-fn map_watch_error(error: notify::Error) -> PyErr {
+fn map_watch_error(error: notify::Error, ignore_permission_denied: bool) -> Option<PyErr> {
```
Comment (samuelcolvin) [inline src/lib.rs:48]: surely better to leave this function unchanged, and just decide what to do with the error where the function is called?

File: src/lib.rs [hunk 2]
```diff
@@ -45,36 +45,48 @@ struct RustNotify {
...
         let mode = if $recursive {
             RecursiveMode::Recursive
         } else {
             RecursiveMode::NonRecursive
         };
         for watch_path in $paths.into_iter() {
-            $watcher.watch(Path::new(&watch_path), mode).map_err(map_watch_error)?;
+            let result = $watcher.watch(Path::new(&watch_path), mode);
+            match result {
+                Err(err) => match map_watch_error(err, $ignore_permission_denied) {
```
Comment (samuelcolvin) [inline src/lib.rs:84]: if here do:

```rust
let err = map_watch_error(err);
if $ignore_permission_denied {
    return err;
}
```

**Review by samuelcolvin — COMMENTED** [review]
and the env var I mentioned.

**Comment by aminalaee** [comment]
Thanks for the review, last time I tried I couldn't figure out how the env vars are configured and used, I will take another look.

File: watchfiles/main.py [hunk 3]
```diff
@@ -319,3 +331,10 @@ def _auto_force_polling() -> bool:
 
     uname = platform.uname()
     return 'microsoft-standard' in uname.release.lower() and uname.system.lower() == 'linux'
+
+
+def _default_ignore_permission_denied(ignore_permission_denied: Optional[bool]) -> bool:
+    if ignore_permission_denied is not None:
+        return ignore_permission_denied
+    env_var = os.getenv('WATCHFILES_IGNORE_PERMISSION_DENIED')
+    return env_var is not None
```
Comment (samuelcolvin) [inline watchfiles/main.py:340]:
```suggestion
    return bool(env_var)
```
To follow the convention that empty strings are equivalent to unsetting the env var.

**Comment by samuelcolvin** [comment]
thanks so much.

**Comment by mshytikov** [comment]
Hi here, thank you for adding the `--ignore-permission-denied`

I just want to flag that I observed strange behaviour with it (Linux). Which might help other people to save some time. 

It stops watching directories if permission-denied occurs on the first directory in the workdir.

Example you have a directory `a`, `b`, `c`:
* if `a` has permission-denied then `workdir` is watched but `a`,  `b` and `c` are not watched anymore - which is unexpected. I would expect `b` and `c` to still be watched. 
* if `b`  has permission-denied then `workdir`, `a`, `c`  are watched and `b` is not - which is expected.

##### Directory setup
``` bash
# LC_COLLATE="C.UTF-8"
# directories setup
mkdir tmp && cd tmp
mkdir a b c
touch a/test.txt b/test.txt c/test.txt
```

##### Example 1 (unexpected behaviour)
```
# set permissions on 'a' directory and make it unaccessible.
sudo chown -R root a b c  && sudo chmod -R go+rwx a b c 
sudo chmod go-rwx a 

# start watching files and send it to background 
watchfiles 'echo change' --ignore-permission-denied &

# change file in workdir and observe the change event 
echo '!' >> test.txt

# change files in 'b' and 'c', no change events observed.
echo '!' >> b/test.txt
echo '!' >> c/test.txt 

# stop watchfiles 
pkill watchfiles
```

##### Example 2 (expected behaviour)
```
# set permissions and make 'a' accessible and 'b' not. 
# change permissions on 'a' directory
sudo chown -R root a b c  && sudo chmod -R go+rwx a b c 
sudo chmod go-rwx b 

# start watching files and send it to background 
watchfiles 'echo change' --ignore-permission-denied &

# change in files in workdir, 'a' and 'c'  and observe all change event 
echo '!' >> test.txt
echo '!' >> a/test.txt
echo '!' >> c/test.txt 

# stop watchfiles 
pkill watchfiles
```

In uvicorn for example [here](https://github.com/encode/uvicorn/issues/1785#issuecomment-1693032220) is recommended to use env var (WATCHFILES_IGNORE_PERMISSION_DENIED) but it does not help because of above.
For my case the only way to avoid the issue while using uvicorn, is to set the `--reload-dir`, but this https://github.com/encode/uvicorn/pull/2107 is not merged yet.

**Pull request merged**